src/transform/src/predicate_pushdown.rs

```diff
@@ -478,33 +483,41 @@ impl PredicatePushdown {
-                    if let Some(list) = get_predicates.remove(&Id::Local(*id)) {
-                        if !list.is_empty() {
-                            // Remove the predicates in `list` from the body.
-                            body.try_visit_mut_post::<_, TransformError>(&mut |e| {
-                                if let MirRelationExpr::Filter { input, predicates } = e {
-                                    if let MirRelationExpr::Get { id: get_id, .. } = **input {
-                                        if get_id == Id::Local(*id) {
-                                            predicates.retain(|p| !list.contains(p));
-                                        }
-                                    }
-                                }
-                                Ok(())
-                            })?;
-                            // Apply the predicates in `list` to value. Canonicalize
-                            // `list` so that plans are always deterministic.
-                            let mut list = list.into_iter().collect::<Vec<_>>();
-                            mz_expr::canonicalize::canonicalize_predicates(
-                                &mut list,
-                                &value.typ().column_types,
-                            );
-                            **value = value.take_dangerous().filter(list);
-                        }
-                    }
+                    Self::push_into_let_binding(get_predicates, id, value, &mut vec![body])?;
 
                     // Continue recursively on the value.
                     self.action(value, get_predicates)
                 }
+                MirRelationExpr::LetRec { ids, values, body } => {
+                    // Note: This could be extended to be able to do a little more pushdowns, see
+                    // https://github.com/MaterializeInc/materialize/issues/18167#issuecomment-1477588262
+
+                    // Pre-compute which Ids are used across iterations
+                    let ids_used_across_iterations = MirRelationExpr::recursive_ids(ids, values)?;
+
+                    // Predicate pushdown within the body
+                    self.action(body, get_predicates)?;
+
+                    // `users` will be the body plus the values of those bindings that we have seen
+                    // so far, while going one-by-one through the list of bindings backwards.
+                    // `users` contains those expressions from which we harvested `get_predicates`,
+                    // and therefore we should attend to all of these expressions when pushing down
+                    // a predicate into a Let binding.
+                    let mut users = vec![&mut **body];
+                    for (id, value) in ids.iter_mut().zip(values).rev() {
+                        // Predicate pushdown from Gets in `users` into the value of a Let binding
+                        if !ids_used_across_iterations.contains(id) {
```

Comment:
I think what you want to characterize is 

> Bindings that are loop-invariant and can be pulled out of the loop to the body

I think `NormalizeLets` does some of this already (if the binding is used only once after that). However, I'm not 100% sure the current characterization as `ids - recursive_ids(ids, values)` is correct.

Reply:
> But what will happen if you have the following structure:

In this situation, the pushdown is excluded by the standard logic of PredicatePushdown (that was already present for normal Gets): An entry in `get_predicates` is the _intersection_ of predicates that are on top of Gets of a particular Id. Here, the `Get y` in the body doesn't have to predicate that the `Get y` in `val_z` has, so the pushdown doesn't happen.

I added one more test to check that the intersecting also correctly includes the body.

Reply:
Oh wait, there might be an issue when pushing something from the body into such a binding that is not in `recursive_ids`. I'm thinking.

Reply:
Thinking through some more cases:
- If in your example only that `Get y` has a predicate which is in the body. In this situation, we don't push into the `val_y`, because this is also correctly handled by the intersecting logic.
- If in your example both `Get y`s have the same predicate. In this situation, we do push into `val_y`, but this is fine.